src/nbl/video/CVKSwapchain.cpp

```diff
@@ -0,0 +1,96 @@
+#include "nbl/video/CVKSwapchain.h"
+
+#include "nbl/video/CVKLogicalDevice.h"
+#include "nbl/video/surface/ISurfaceVK.h"
+#include "nbl/video/CVulkanImage.h"
+
+namespace nbl {
+namespace video
+{
+
+CVKSwapchain::CVKSwapchain(SCreationParams&& params, CVKLogicalDevice* dev) : ISwapchain(std::move(params)), m_device(dev)
+{
+    VkSwapchainCreateInfoKHR ci;
+    ci.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
+    ci.pNext = nullptr;
+    ci.minImageCount = params.minImageCount;
+    ci.clipped = VK_TRUE;
+    ci.imageArrayLayers = params.arrayLayers;
+    // TODO function mapping ISurface::SFormat -> VkColorSpaceKHR
+    //ci.imageColorSpace = ...
+    ci.presentMode = static_cast<VkPresentModeKHR>(params.presentMode);
+    ci.imageExtent = { params.width, params.height };
+    ci.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE; // single queue famility at a time
+    ci.queueFamilyIndexCount = params.queueFamilyIndices->size();
+    ci.pQueueFamilyIndices = params.queueFamilyIndices->data();
+    ci.preTransform = VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR;
+    ci.oldSwapchain = VK_NULL_HANDLE;
+    //ci.imageUsage = ... // TODO (we dont have the enum yet)
+    ci.flags = 0;
+    ci.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
+    ci.imageFormat = static_cast<VkFormat>(params.surfaceFormat.format);
+    ci.surface = static_cast<ISurfaceVK*>(params.surface.get())->getInternalObject();
```

Comment:
need that `castDown` function for safety, an if the castDown is going to occur in a `CVKSwapchain` method then you need to move it out of the constructor into a `create` static method